### PR TITLE
ENH: Logistic completeness function

### LINKED ISSUE #519: ENH: Logistic Completeness function
## Description
This function returns the logistic completeness function:

                  1
    p(m) = -----------------
           1 + exp{-k(m-m0)}

that describes the probability `p(m)` that a galaxy of magnitude `m` is detected in the band.

See also the discussion around eq. (7) and fig. 2 in the ALHAMBRA paper by Lopez-Sanjuan et al. (2017).

## Inputs
 - apparent magnitudes
 - 95% completeness
 - 50% completeness

## Outputs
 - Logistic completeness function

## References
- [Lopez-Sanjuan et al. 2017](https://www.aanda.org/articles/aa/full_html/2017/03/aa29517-16/aa29517-16.html)

### PR DESCRIPTION
**Pull request by philipp128**
## Description
This PR adds a feature to calculate the logistic completeness function. Closes #519.

## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/main/CONTRIBUTING.rst)
- [x] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

### CODE REVIEW

File: skypy/utils/completeness_function.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+def logistic_completeness_function(m, m95, m50):
+    r'''Compute logistic completeness function.
+
+    This function calculates the logistic completeness function
+
+    .. math::
+
+        p(m) = \frac{1}{1 + \exp\left[\frac{\ln(1/19)}{m_{95} - m_{50}} (m - m_{50})\right]}\;.
+
```
Comment (Lucia-Fonseca) [inline skypy/utils/completeness_function.py:24]:
```suggestion
    .. math::

        p(m) = \frac{1}{1 + e^{\kappa (m - m_{50})}}
        
     with :math:`\kappa = \frac{\ln(1/19)}{m_{95} - m_{50}}`.

```
You could try make it more readable maybe?

File: skypy/utils/completeness_function.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    m50 : array_like
+        50% completeness magnitude.
+    Returns
+    -------
+    p : array like
+        Result of the completeness function.
+
+    '''
+
+    return 1. / (1 + np.exp(np.log(1. / 19) / np.subtract(m95, m50) * np.subtract(m, m50)))
```
Comment (Lucia-Fonseca) [inline skypy/utils/completeness_function.py:43]:
```suggestion
    kappa = np.log(1. / 19) / np.subtract(m95, m50)
    return 1. / (1 + np.exp( kappa * np.subtract(m, m50)))
```

File: skypy/utils/completeness_function.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    .. math::
+
+        p(m) = \frac{1}{1 + \exp\left[\frac{\ln(1/19)}{m_{95} - m_{50}} (m - m_{50})\right]}\;.
+
+    which describes the probability :math:`p(m)` that an object of magnitude :math:`m` is detected
+    in the band.
+
+    Parameters
+    ----------
+    m : array_like
```
Comment (Lucia-Fonseca) [inline skypy/utils/completeness_function.py:30]: Is this an array? or a n-d array? I'm guessing from your test.

Please add more description to the parameters section

File: skypy/utils/completeness_function.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        Magnitudes. Can be multidimensional for computing with multiple filter bands.
+    m95 : array_like
+        95% completeness magnitude.
+    m50 : array_like
+        50% completeness magnitude.
+    Returns
+    -------
+    p : array like
```
Comment (Lucia-Fonseca) [inline skypy/utils/completeness_function.py:38]:
```suggestion
    probability : array like
```

File: skypy/utils/completeness_function.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        Result of the completeness function.
```
Comment (Lucia-Fonseca) [inline skypy/utils/completeness_function.py:39]:
```suggestion
        Probability of detecting a galaxy following with magnitude :math:`m`. 
```
Or something like that. You need to also specify the shape.

File: skypy/utils/tests/test_completeness_function.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import numpy as np
+
+
+def test_logistic_completeness_function():
+    from skypy.utils.completeness_function import logistic_completeness_function
+
+    # Test that (nb, nm) array is returned if m in (mb, nm) and m95 and m50 in (nb, ) are given
+    m = np.ndarray((5, 50))
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:8]: Is this for 5 galaxies with 50 values of magnitudes?

File: skypy/utils/tests/test_completeness_function.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    m[:] = np.linspace(18, 28, 50)
+    m = m.T
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:10]: `np.reshape` is much better for this
https://numpy.org/doc/stable/reference/generated/numpy.reshape.html

File: skypy/utils/tests/test_completeness_function.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    # Test that returned shape is equal to shape of m if it is 2D array and m05 and m50 floats
+    # array
+    m = np.ndarray((5, 50))
+    m[:] = np.linspace(18, 28, 50)
+    m = m.T
+    m50 = 24.7
+    m95 = 24
+    p = logistic_completeness_function(m, m95, m50)
+    assert m.shape == p.shape
+
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:32]: Have you checked other corner cases? For example, m50, m95 being arrays and m scalar? If this example does not make physical sense, you should specify it in the docstring.

File: skypy/utils/tests/test_completeness_function.py (hunk 6 above)
Comment (Lucia-Fonseca) [reply skypy/utils/tests/test_completeness_function.py:8]: also use unique variable names for each test block. You could use here `m_ndarray`

File: skypy/utils/tests/test_completeness_function.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+def test_logistic_completeness_function():
+    from skypy.utils.completeness_function import logistic_completeness_function
+
+    # Test that (nb, nm) array is returned if m in (mb, nm) and m95 and m50 in (nb, ) are given
+    m = np.ndarray((5, 50))
+    m[:] = np.linspace(18, 28, 50)
+    m = m.T
+    m50 = np.array([24.7, 25, 23, 23.5, 28])
+    m95 = np.array([24, 23, 22, 22.3, 25])
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:12]:
```suggestion
    m50_array = np.array([24.7, 25, 23, 23.5, 28])
    m95_array = np.array([24, 23, 22, 22.3, 25])
```

File: skypy/utils/tests/test_completeness_function.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    p = logistic_completeness_function(m, m95, m50)
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:13]:
```suggestion
    prob_ndarray = logistic_completeness_function(m_ndarray, m95_array, m50_array)
```

File: skypy/utils/tests/test_completeness_function.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    assert m.shape == p.shape
+
+    # Test that returned shape is equal to shape of m if it is 1D array and m05 and m50 floats
+    m = np.linspace(18, 28, 50)
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:17]:
```suggestion
    m_1darray = np.linspace(18, 28, 50)
```

File: skypy/utils/tests/test_completeness_function.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    m50 = 24.7
+    m95 = 24
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:19]:
```suggestion
    m50_scalar = 24.7
    m95_scalar = 24
```

File: skypy/utils/tests/test_completeness_function.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    p = logistic_completeness_function(m, m95, m50)
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:20]:
```suggestion
    prob_1darray = logistic_completeness_function(m_1darray, m95_scalar, m50_scalar)
```

File: skypy/utils/tests/test_completeness_function.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    assert m.shape == p.shape
+
+    # Test that returned shape is equal to shape of m if it is 2D array and m05 and m50 floats
+    # array
+    m = np.ndarray((5, 50))
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:25]: same with the variable names

File: skypy/utils/tests/test_completeness_function.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    m[:] = np.linspace(18, 28, 50)
+    m = m.T
+    m50 = 24.7
+    m95 = 24
```
Comment (Lucia-Fonseca) [inline skypy/utils/tests/test_completeness_function.py:29]: if you named it correctly, you don't need to re-define it here, it's a waste of memory.

**Review by Lucia-Fonseca — CHANGES_REQUESTED** [review]
Comments on readability, naming convention and potentially extra unit tests.

File: skypy/utils/completeness_function.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
+"""Completeness functions.
```
[the remark this answers is not in the document]
Comment (philipp128) [reply skypy/utils/completeness_function.py:1]: I disagree. In `skypy.utils.special` we also wrote "Special functions". I wanted to be consistent. There might be other completeness functions in the future.

I know that two function are defined in `skypy.utils.special` but only one is visible to the average user in the docs and the other one only a "help" function.

File: skypy/utils/completeness_function.py (hunk 3 above)
Comment (philipp128) [reply skypy/utils/completeness_function.py:30]: It is an `array_like`. The function is not restricted to arrays. You can also use scalars, lists etc.

File: skypy/utils/tests/test_completeness_function.py (hunk 6 above)
Comment (philipp128) [reply skypy/utils/tests/test_completeness_function.py:8]: That's is 50 values of magnitudes for 5 cases with different completeness values.

I personally dislike giving unique variable names. I prefer redefining the parameters to save memory.

File: skypy/utils/tests/test_completeness_function.py (hunk 15 above)
Comment (philipp128) [reply skypy/utils/tests/test_completeness_function.py:29]: It is the opposite of a waste of memory, no?

**Comment by philipp128** [comment]
Docs are failing. I am waiting to resolve this right now cause I have a problem with the docs anyway. 

The docs for completeness function is not rendered properly. It is included but not linked or specified -> see attached image. I do not know where the problem is.

@rrjbca do you know whether I forgot sth?

image

**Comment by rrjbca** [comment]
I would also suggest making this part of skypy.utils.photometry. In effect the completeness is an extrinsic photometric property of the sources. This would again save creating a new submodule.

File: skypy/utils/completeness_function.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        is array_like of shape (nb, ) it returns array_like of shape (nb, ).
+
+    References
+    -----------
+    .. [1] Lopez-Sanjuan C. et al., 2017, A&A, 599, A62
+
+    '''
+
+    kappa = np.log(1. / 19) / np.subtract(magnitude_95, magnitude_50)
+    return 1. / (1 + np.exp(kappa * np.subtract(magnitude, magnitude_50)))
```
Comment (rrjbca) [inline skypy/utils/completeness_function.py:59]: [`scipy.special.expit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.expit.html)

File: skypy/utils/completeness_function.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/utils/completeness_function.py]: > I know that two function are defined in skypy.utils.special but only one is visible to the average user in the docs and the other one only a "help" function.

Which one is hidden?

File: skypy/utils/photometry.py [hunk 18]
```diff
@@ -392,3 +394,53 @@ def magnitude_error_rykoff(magnitude, magnitude_limit, magnitude_zp, a, b, error
     error = 2.5 / np.log(10) * np.sqrt((1 + flux_noise / flux) / (flux * t_eff))
 
     return np.minimum(error, error_limit)
+
+
+def logistic_completeness_function(magnitude, magnitude_95, magnitude_50):
+    r'''Compute logistic completeness function.
```
Comment (Lucia-Fonseca) [inline skypy/utils/photometry.py:400]: This should be more of a title, either
```suggestion
    r'''Logistic completeness function.
```

or
```suggestion
    r'''Computation of the logistic completeness function.
```

**Review by Lucia-Fonseca — DISMISSED** [review]
Just a comment on docstrings, but looks good.

File: skypy/utils/photometry.py (hunk 18 above)
Comment (philipp128) [reply skypy/utils/photometry.py]: I used your first suggestion and changed the docstring.

I guess now we need to see if @rrjbca agrees with the changes.

**Review by rrjbca — CHANGES_REQUESTED** [review]
- You shouldn't need six tests for different combinations of array shapes; one slightly more complex broadcasting test should be sufficient e.g:
```
m = np.full((2, 1, 1), 21)
m95 = np.full((3, 1), 22)
m50 = np.full(5, 23)
p = logistic_completeness_function(m, m95, m50)
assert p.shape = np.broadcast(m, m95, m50).shape
```

- In your final test check more magnitudes that have analytic values:
```
m95 = 24
m50 = 25
m = [np.finfo(np.float64).min, m95, m50, 2*m50-m95, np.finfo(np.float64).max]
p = logistic_completeness_function(m, m95, m50)
assert np.allclose(p, [1, 0.95, 0.5, 0.05, 0])
```

- The reference to López-Sanjuan 2017 is not linked anywhere in the docstring. Also note the correct spelling of the name with the accent.

File: skypy/utils/tests/test_photometry.py [hunk 19]
```diff
@@ -212,3 +212,21 @@ def test_magnitude_error_rykoff():
     error_limit = 1
     error = magnitude_error_rykoff(magnitude, magnitude_limit, magnitude_zp, a, b, error_limit)
     assert error == error_limit
+
+
+def test_logistic_completeness_function():
+    from skypy.utils.photometry import logistic_completeness_function
+
+    # Test that correct shapeis returned
```
Comment (rrjbca) [inline skypy/utils/tests/test_photometry.py:220]:
```suggestion
    # Test that arguments broadcast correctly
```

**Review by rrjbca — APPROVED** [review]
